arrows/core/render_mesh_depth_map.cxx

```diff
@@ -0,0 +1,213 @@
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief Header for core render_mesh_depth_map function
+ */
+
+#include "render_mesh_depth_map.h"
+
+#include <vital/types/camera_perspective.h>
+#include <vital/types/image.h>
+#include <vital/types/vector.h>
+
+
+namespace {
+
+/// Helper function to check if a point is inside a triangle
+bool is_point_inside_triangle(const kwiver::vital::vector_2d& p,
+                              const kwiver::vital::vector_2d& a,
+                              const kwiver::vital::vector_2d& b,
+                              const kwiver::vital::vector_2d& c)
+{
+  kwiver::vital::vector_2d AB = b - a;
+  kwiver::vital::vector_2d AC = c - a;
+  kwiver::vital::vector_2d AP = p - a;
+  double inv_total_area = 1.0 / (AB[0] * AC[1] - AB[1] * AC[0]);
+  double area_1 = inv_total_area * (AB[0] * AP[1] - AB[1] * AP[0]);
+  double area_2 = inv_total_area * (AP[0] * AC[1] - AP[1] * AC[0]);
+  return area_1 >= 0 && area_2 >= 0 && (area_1 + area_2) <= 1;
+}
+
+/// Helper function to compute barycentric coordinates of p w.r.t the triangle abc
+kwiver::vital::vector_3d barycentric_coordinates(const kwiver::vital::vector_2d& p,
+                                                 const kwiver::vital::vector_2d& a,
+                                                 const kwiver::vital::vector_2d& b,
+                                                 const kwiver::vital::vector_2d& c)
+{
+  kwiver::vital::matrix_3x3d abc;
+  abc << a, b, c, 1, 1, 1;
+  double det_inv = 1.0 / abc.determinant();
+  kwiver::vital::vector_3d res;
+  res(0) = ((b(1) - c(1)) * (p(0) - c(0)) - (b(0) - c(0)) * (p(1) - c(1))) * det_inv;
+  res(1) = ((c(1) - a(1)) * (p(0) - c(0)) - (c(0) - a(0)) * (p(1) - c(1))) * det_inv;
+  res(2) = 1.0 - res(0) - res(1);
+  return res;
+}
+
+}
+
+namespace kwiver {
+namespace arrows {
+
+
+vital::image_container_sptr render_mesh_depth_map(vital::mesh_sptr mesh, vital::camera_perspective_sptr camera)
+{
+  unsigned int nb_vertices = mesh->num_verts();
+  vital::mesh_vertex_array<3>& vertices = dynamic_cast< vital::mesh_vertex_array<3>& >(mesh->vertices());
```

Comment:
You could try this here to see if it helps with the cast failure on the dashboard
```c++
vital::mesh_vertex_array<3>& vertices = mesh->template vertices<3>();
```

Reply:
The bad cast is not caused by this line, but by the dynamic_cast on faces (line 113)
`dynamic_cast< const vital::mesh_regular_face_array<3>& >(mesh->faces())`